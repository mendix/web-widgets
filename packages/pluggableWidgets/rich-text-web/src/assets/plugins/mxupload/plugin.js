// Traverse through the content definition and attach filebrowser to
// elements with 'filebrowser' attribute.
//
// @param String
//            dialogName Dialog name.
// @param {CKEDITOR.dialog.definitionObject}
//            definition Dialog definition.
// @param {Array}
//            elements Array of {@link CKEDITOR.dialog.definition.content}
//            objects.
function attachFileBrowser(editor, dialogName, definition, elements) {
    if (!elements || !elements.length) return;

    let element;

    for (let i = elements.length; i--; ) {
        element = elements[i];

        if (element.type == "hbox" || element.type == "vbox" || element.type == "fieldset")
            attachFileBrowser(editor, dialogName, definition, element.children);

        if (!element.filebrowser) continue;

        if (typeof element.filebrowser == "string") {
            const fb = {
                action: "QuickUpload",
                target: element.filebrowser
            };
            element.filebrowser = fb;
        }

        if (element.filebrowser.action == "QuickUpload" && element.type === "file") {
            element.onChange = function (evt) {
                const fileInput = this.getInputElement();
                const file = fileInput.$.files[0];
                if (file.size > editor.uploadImageMaxSize * 1024) {
                    alert(`Image is too large. Select an image less than ${editor.uploadImageMaxSize} KB`);
                    return;
                }

                const dialog = evt.sender.getDialog();
                editor._.filebrowserSe = evt.sender;

                window
                    .fetch(editor.uploadImageEndpoint, {
                        method: "POST",
                        body: file,
                        headers: {
                            Origin: window.location.origin,
                            "Content-Type": "text/html",
                            "x-csrf-token": window.mx ? window.mx.session.getConfig("csrftoken") : ""
                        }
                    })
                    .then(response => {
                        if (!response.ok || response.status >= 400) {
                            throw Error(
                                JSON.stringify({
                                    status: response.status,
                                    statusText: response.statusText,
                                    ok: response.ok,
                                    body: response.body
                                })
                            );
                        }
                        return response.text();
                    })
                    .then(imagePath => setUrl(editor, dialog, imagePath, file.name));
                return false;
            };
        }
        if (element.filebrowser.action == "QuickUpload" && element["for"]) {
            element.hidden = true;
            const params = element.filebrowser.params || {};
            params.CKEditor = editor.name;
            params.CKEditorFuncNum = editor._.filebrowserFn;
            if (!params.langCode) params.langCode = editor.langCode;
            const fileInput = definition.getContents(element["for"][0]).get(element["for"][1]);
            fileInput.filebrowser = element.filebrowser;
        }
    }
}

function setUrl(editor, dialog, fileUrl, imageName) {
    const targetInput = editor._.filebrowserSe["for"];

    if (targetInput) dialog.getContentElement(targetInput[0], targetInput[1]).reset();

    if (fileUrl) updateTargetElement(fileUrl, editor._.filebrowserSe, dialog, imageName);
}

function updateTargetElement(url, sourceElement, dialog, imageName) {
    const targetElement = sourceElement.filebrowser.target || null;

    // If there is a reference to targetElement, update it.
    if (targetElement) {
        const target = targetElement.split(":");
        const element = dialog.getContentElement(target[0], target[1]);
        if (element) {
            element.setValue(url);
            dialog.selectPage(target[0]);
        }
        const altElement = dialog.getContentElement(target[0], "txtAlt");
        if (altElement) {
            altElement.setValue(imageName);
        }
    }
}

// Returns true if filebrowser is configured in one of the elements.
//
// @param {CKEDITOR.dialog.definitionObject}
//            definition Dialog definition.
// @param String
//            tabId The tab id where element(s) can be found.
// @param String
//            elementId The element id (or ids, separated with a semicolon) to check.
function isConfigured(definition, tabId, elementId) {
    if (elementId.indexOf(";") !== -1) {
        const ids = elementId.split(";");
        for (let i = 0; i < ids.length; i++) {
            if (isConfigured(definition, tabId, ids[i])) return true;
        }
        return false;
    }

    const elementFileBrowser = definition.getContents(tabId).get(elementId).filebrowser;
    return !!elementFileBrowser;
}

CKEDITOR.plugins.add("mxupload", {
    init: function (editor) {
        editor._.filebrowserFn = CKEDITOR.tools.addFunction(setUrl, editor);
        editor.on("destroy", function () {
            CKEDITOR.tools.removeFunction(this._.filebrowserFn);
        });
    }
});

CKEDITOR.on("dialogDefinition", function (evt) {
    if (!evt.editor.plugins.mxupload) return;

    let definition = evt.data.definition,
        element;
    // Associate mxupload to elements with 'filebrowser' attribute.
    for (let i = 0; i < definition.contents.length; ++i) {
        if ((element = definition.contents[i])) {
            attachFileBrowser(evt.editor, evt.data.name, definition, element.elements);
            if (element.hidden && element.filebrowser)
                element.hidden = !isConfigured(definition, element.id, element.filebrowser);
        }
    }
});
