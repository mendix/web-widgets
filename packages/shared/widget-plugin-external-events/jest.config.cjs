module.exports = {
    modulePathIgnorePatterns: ["<rootDir>/dist/"],
    transform: {
        "^.+\\.(t|j)sx?$": [
            "@swc/jest",
            {
                jsc: {
                    transform: {
                        react: {
                            runtime: "automatic"
                        }
                    }
                }
            }
        ]
    },
    testEnvironment: "jsdom",
    moduleNameMapper: {
        "(.+)\\.js": "$1"
    },
    extensionsToTreatAsEsm: [".ts", ".tsx"],
    transformIgnorePatterns: ["node_modules/(?!nanoevents)/"]
};
