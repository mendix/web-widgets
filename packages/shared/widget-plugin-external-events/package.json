{
  "name": "@mendix/widget-plugin-external-events",
  "version": "0.1.0",
  "description": "Plugin for subscribing to and publishing widget events",
  "license": "Apache-2.0",
  "copyright": "Â© Mendix Technology BV 2023. All rights reserved.",
  "repository": {
    "type": "git",
    "url": "https://github.com/mendix/web-widgets.git"
  },
  "type": "module",
  "exports": {
    "./*": "./dist/*.js"
  },
  "typesVersions": {
    "*": {
      "*": [
        "dist/*"
      ]
    }
  },
  "files": [
    "dist",
    "!*.map"
  ],
  "scripts": {
    "build": "tsc -p tsconfig.build.json",
    "dev": "tsc -p tsconfig.build.json --watch",
    "format": "prettier --write .",
    "lint": "eslint --ext .ts,.tsx ./src",
    "prepare": "pnpm build",
    "test": "jest"
  },
  "devDependencies": {
    "@mendix/eslint-config-web-widgets": "workspace:*",
    "@mendix/prettier-config-web-widgets": "workspace:*",
    "@mendix/tsconfig-web-widgets": "workspace:*",
    "@swc/core": "^1.3.37",
    "@swc/jest": "^0.2.24",
    "jest-environment-jsdom": "^29.4.1"
  },
  "dependencies": {
    "nanoevents": "^9.0.0"
  }
}
