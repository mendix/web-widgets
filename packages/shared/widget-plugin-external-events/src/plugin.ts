import { createNanoEvents, Emitter, Unsubscribe } from "nanoevents";

const pluginPathExternalEvents = "com.mendix.widgets.web.plugin.externalEvents" as const;

export type Listener = (...args: any[]) => void;
export type Emit = (channelName: string, eventName: string, ...args: any[]) => void;

declare global {
    interface Window {
        [pluginPathExternalEvents]?: PluginExternalEvents;
    }
}

export interface PluginExternalEvents {
    emit: Emit;
    subscribe(channelName: string, eventName: string, callback: Listener): Unsubscribe;
    unsubscribe(channelName: string, eventName: string, callback: Listener): void;
}

class ExternalEvents implements PluginExternalEvents {
    private channels: Map<string, Emitter>;
    readonly id = Math.random().toString().slice(2);

    constructor() {
        this.channels = new Map<string, Emitter>();
    }

    private disposeChannel(channelName: string): boolean {
        const emitter = this.channels.get(channelName);
        if (emitter) {
            emitter.events = {};
        }

        return this.channels.delete(channelName);
    }

    private getChannel(channelName: string): Emitter {
        let emitter = this.channels.get(channelName);

        if (emitter) {
            return emitter;
        }

        emitter = createNanoEvents();
        this.channels.set(channelName, emitter);
        return emitter;
    }

    emit(channelName: string, eventName: string, ...args: any[]): void {
        this.channels.get(channelName)?.emit(eventName, ...args);
    }

    subscribe(channelName: string, eventName: string, listener: Listener): Unsubscribe {
        this.getChannel(channelName).on(eventName, listener);
        return () => this.unsubscribe(channelName, eventName, listener);
    }

    unsubscribe(channelName: string, eventName: string, listener: Listener): void {
        const channel = this.channels.get(channelName);
        if (!channel) {
            return;
        }

        const listeners = channel.events[eventName];
        if (listeners) {
            channel.events[eventName] = listeners.filter(fn => fn !== listener);
        }
        const listenersTotal = Object.values(channel.events)
            .flat()
            .reduce((acc, fn) => (fn ? acc + 1 : acc), 0);

        if (listenersTotal === 0) {
            this.disposeChannel(channelName);
        }
    }
}

function pluginSetup(): PluginExternalEvents {
    if (Object.prototype.hasOwnProperty.call(window, pluginPathExternalEvents)) {
        throw new Error("Widget plugin external events: plugin already initialized!");
    }

    return new ExternalEvents();
}

function requirePlugin(): PluginExternalEvents {
    return (window[pluginPathExternalEvents] ??= pluginSetup());
}

function deletePlugin(): void {
    delete window[pluginPathExternalEvents];
}

export { requirePlugin, deletePlugin };
