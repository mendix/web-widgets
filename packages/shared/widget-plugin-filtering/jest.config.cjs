module.exports = {
    modulePathIgnorePatterns: ["<rootDir>/dist/"],
    transform: {
        "^.+\\.(t|j)sx?$": [
            "@swc/jest",
            {
                jsc: {
                    transform: {
                        react: {
                            runtime: "automatic"
                        }
                    }
                }
            }
        ]
    },
    moduleDirectories: ["node_modules", "src"],
    moduleNameMapper: {
        "(.+)\\.js": "$1"
    },
    extensionsToTreatAsEsm: [".ts"],
    testEnvironment: "jsdom",
    collectCoverage: !process.env.CI,
    coverageProvider: "v8"
};
