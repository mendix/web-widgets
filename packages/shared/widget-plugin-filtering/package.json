{
  "name": "@mendix/widget-plugin-filtering",
  "version": "1.0.0",
  "description": "Filtering API plugin.",
  "license": "Apache-2.0",
  "copyright": "Â© Mendix Technology BV 2023. All rights reserved.",
  "repository": {
    "type": "git",
    "url": "https://github.com/mendix/web-widgets.git"
  },
  "type": "module",
  "main": "./dist/main.js",
  "types": "./dist/main.d.ts",
  "exports": {
    ".": "./dist/main.js",
    "./controls": "./dist/controls.js"
  },
  "typesVersions": {
    "*": {
      "controls": [
        "./dist/controls.d.ts"
      ]
    }
  },
  "files": [
    "dist",
    "!*.map"
  ],
  "scripts": {
    "compile": "tsc",
    "format": "prettier --write .",
    "lint": "eslint --ext .ts,.tsx ./src",
    "prepare": "tsc",
    "dev": "tsc --watch"
  },
  "devDependencies": {
    "@mendix/eslint-config-web-widgets": "workspace:*",
    "@mendix/prettier-config-web-widgets": "workspace:*",
    "@mendix/tsconfig-web-widgets": "workspace:*"
  },
  "dependencies": {
    "@mendix/widget-plugin-hooks": "workspace:*",
    "@mendix/widget-plugin-platform": "workspace:*",
    "@mendix/widget-plugin-external-events": "workspace:*",
    "@mendix/widget-plugin-filter-selector": "workspace:*"
  }
}
