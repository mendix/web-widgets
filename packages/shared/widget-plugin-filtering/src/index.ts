export * from "./custom-filter-api/DropdownStoreProvider";
