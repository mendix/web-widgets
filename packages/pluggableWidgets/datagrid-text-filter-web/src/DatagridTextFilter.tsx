import { createElement, ReactElement, useRef } from "react";
import { DatagridTextFilterContainerProps, DefaultFilterEnum } from "../typings/DatagridTextFilterProps";

import { FilterComponent } from "./components/FilterComponent";
import { Alert } from "@mendix/widget-plugin-component-kit/Alert";
import { generateUUID } from "@mendix/widget-plugin-platform/framework/generate-uuid";
import { FilterType, getFilterDispatcher, useDefaultValue } from "@mendix/widget-plugin-filtering";

import {
    attribute,
    contains,
    endsWith,
    equals,
    greaterThan,
    greaterThanOrEqual,
    lessThan,
    lessThanOrEqual,
    literal,
    notEqual,
    or,
    startsWith
} from "mendix/filters/builders";
import { FilterCondition } from "mendix/filters";
import { ListAttributeValue } from "mendix";
import { translateFilters } from "./utils/filters";

export default function DatagridTextFilter(props: DatagridTextFilterContainerProps): ReactElement {
    const id = useRef(`TextFilter${generateUUID()}`);
    const defaultValue = useDefaultValue(props.defaultValue);

    const FilterContext = getFilterDispatcher();
    const alertMessage = (
        <Alert bootstrapStyle="danger">
            The Text filter widget must be placed inside the header of the Data grid 2.0 or Gallery widget.
        </Alert>
    );
    const alertMessageMultipleFilters = (
        <Alert bootstrapStyle="danger">
            The Text filter widget can&apos;t be used with the filters options you have selected. It requires a
            &quot;Hashed string or String&quot; attribute to be selected.
        </Alert>
    );

    return FilterContext?.Consumer ? (
        <FilterContext.Consumer>
            {filterContextValue => {
                if (
                    !filterContextValue ||
                    !filterContextValue.filterDispatcher ||
                    (!filterContextValue.singleAttribute && !filterContextValue.multipleAttributes)
                ) {
                    return alertMessage;
                }
                const {
                    filterDispatcher,
                    singleAttribute,
                    multipleAttributes,
                    singleInitialFilter,
                    multipleInitialFilters
                } = filterContextValue;

                const attributes = [
                    ...(singleAttribute ? [singleAttribute] : []),
                    ...(multipleAttributes ? findAttributesByType(multipleAttributes) ?? [] : [])
                ];

                if (attributes.length === 0) {
                    if (multipleAttributes) {
                        return alertMessageMultipleFilters;
                    }
                    return alertMessage;
                }

                const parentFilter = singleInitialFilter
                    ? translateFilters(singleInitialFilter)
                    : translateFilters(multipleInitialFilters?.[attributes[0].id]);

                const errorMessage = getAttributeTypeErrorMessage(attributes[0].type);
                if (errorMessage) {
                    return <Alert bootstrapStyle="danger">{errorMessage}</Alert>;
                }

                if (defaultValue === null) {
                    return null;
                }

                return (
                    <FilterComponent
                        adjustable={props.adjustable}
                        className={props.class}
                        defaultFilter={parentFilter?.type ?? props.defaultFilter}
                        value={parentFilter?.value ?? defaultValue}
                        changeDelay={props.delay}
                        id={id.current}
                        placeholder={props.placeholder?.value}
                        screenReaderButtonCaption={props.screenReaderButtonCaption?.value}
                        screenReaderInputCaption={props.screenReaderInputCaption?.value}
                        styles={props.style}
                        tabIndex={props.tabIndex}
                        parentChannelName={filterContextValue.eventsChannelName ?? null}
                        name={props.name}
                        onChange={(value: string, type: DefaultFilterEnum): void => {
                            props.valueAttribute?.setValue(value);
                            props.onChange?.execute();
                            const conditions = attributes
                                ?.map(attribute => getFilterCondition(attribute, value, type))
                                .filter((filter): filter is FilterCondition => filter !== undefined);
                            filterDispatcher({
                                getFilterCondition: () =>
                                    conditions && conditions.length > 1 ? or(...conditions) : conditions?.[0],
                                filterType: FilterType.STRING
                            });
                        }}
                    />
                );
            }}
        </FilterContext.Consumer>
    ) : (
        alertMessage
    );
}

function findAttributesByType(multipleAttributes?: {
    [key: string]: ListAttributeValue;
}): ListAttributeValue[] | undefined {
    if (!multipleAttributes) {
        return undefined;
    }
    return Object.keys(multipleAttributes)
        .map(key => multipleAttributes[key])
        .filter(attr => attr.type.match(/HashString|String/));
}

function getAttributeTypeErrorMessage(type?: string): string | null {
    return type && !type.match(/HashString|String/)
        ? "The attribute type being used for Text filter is not 'Hashed string or String'"
        : null;
}

function getFilterCondition(
    listAttribute: ListAttributeValue,
    value: string,
    type: DefaultFilterEnum
): FilterCondition | undefined {
    if (!listAttribute || !listAttribute.filterable || (type !== "empty" && type !== "notEmpty" && !value)) {
        return undefined;
    }

    const filters = {
        contains,
        startsWith,
        endsWith,
        greater: greaterThan,
        greaterEqual: greaterThanOrEqual,
        equal: equals,
        notEqual,
        smaller: lessThan,
        smallerEqual: lessThanOrEqual,
        empty: equals,
        notEmpty: notEqual
    };

    return filters[type](
        attribute(listAttribute.id),
        literal(type === "empty" || type === "notEmpty" ? undefined : value)
    );
}
