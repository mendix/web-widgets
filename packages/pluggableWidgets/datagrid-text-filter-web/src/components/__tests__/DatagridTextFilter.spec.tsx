import "@testing-library/jest-dom";
import { FilterContextValue } from "@mendix/widget-plugin-filtering";
import {
    actionValue,
    dynamicValue,
    EditableValueBuilder,
    ListAttributeValueBuilder
} from "@mendix/widget-plugin-test-utils";
import { render, screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createContext, createElement } from "react";
import DatagridTextFilter from "../../DatagridTextFilter";
import { requirePlugin, deletePlugin } from "@mendix/widget-plugin-external-events/plugin";

const commonProps = {
    class: "filter-custom-class",
    tabIndex: 0,
    name: "filter-test",
    defaultFilter: "equal" as const,
    adjustable: true,
    advanced: false,
    delay: 1000
};

jest.useFakeTimers();

describe("Text Filter", () => {
    describe("with single instance", () => {
        afterEach(() => {
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });

        describe("with defaultValue prop", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder().withType("String").withFilterable(true).build()
                } as FilterContextValue);
            });

            it("don't sync value when defaultValue changes from undefined to string", async () => {
                const { rerender } = render(<DatagridTextFilter {...commonProps} defaultValue={undefined} />);

                expect(screen.getByRole("textbox")).toHaveValue("");

                // rerender component with new `defaultValue`
                const defaultValue = dynamicValue<string>("xyz");
                rerender(<DatagridTextFilter {...commonProps} defaultValue={defaultValue} />);
                expect(screen.getByRole("textbox")).toHaveValue("");
            });

            it("don't sync value when defaultValue changes from string to string", async () => {
                const { rerender } = render(
                    <DatagridTextFilter {...commonProps} defaultValue={dynamicValue<string>("abc")} />
                );

                expect(screen.getByRole("textbox")).toHaveValue("abc");

                // rerender component with new `defaultValue`
                const defaultValue = dynamicValue<string>("xyz");
                rerender(<DatagridTextFilter {...commonProps} defaultValue={defaultValue} />);
                expect(screen.getByRole("textbox")).toHaveValue("abc");
            });

            it("don't sync value when defaultValue changes from string to undefined", async () => {
                const { rerender } = render(
                    <DatagridTextFilter {...commonProps} defaultValue={dynamicValue<string>("abc")} />
                );

                expect(screen.getByRole("textbox")).toHaveValue("abc");

                // rerender component with new `defaultValue`
                rerender(<DatagridTextFilter {...commonProps} defaultValue={undefined} />);
                expect(screen.getByRole("textbox")).toHaveValue("abc");
            });
        });

        describe("with single attribute", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder().withType("String").withFilterable(true).build()
                } as FilterContextValue);
            });

            it("renders correctly", () => {
                const { asFragment } = render(<DatagridTextFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            it("triggers attribute and onchange action on change filter value", async () => {
                const action = actionValue();
                const attribute = new EditableValueBuilder<string>().build();
                render(<DatagridTextFilter {...commonProps} onChange={action} valueAttribute={attribute} />);

                const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
                await user.type(screen.getByRole("textbox"), "B");

                jest.runOnlyPendingTimers();

                expect(attribute.setValue).toHaveBeenCalled();
                expect(action.execute).toHaveBeenCalled();
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });
        });

        describe("with multiple attributes", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    multipleAttributes: {
                        attribute1: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withType("String")
                            .withFilterable(true)
                            .build(),
                        attribute2: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withType("HashString")
                            .withFilterable(true)
                            .build()
                    }
                } as FilterContextValue);
            });

            it("renders correctly", () => {
                const { asFragment } = render(<DatagridTextFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });
        });

        describe("with wrong attribute's type", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder().withType("Decimal").withFilterable(true).build()
                } as FilterContextValue);
            });

            it("renders error message", () => {
                const { asFragment } = render(<DatagridTextFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });
        });

        describe("with wrong multiple attributes' types", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    multipleAttributes: {
                        attribute1: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withType("Decimal")
                            .withFilterable(true)
                            .build(),
                        attribute2: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withType("Long")
                            .withFilterable(true)
                            .build()
                    }
                } as FilterContextValue);
            });

            it("renders error message", () => {
                const { asFragment } = render(<DatagridTextFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });
        });

        describe("with no context", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });

            it("renders error message", () => {
                const { asFragment } = render(<DatagridTextFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });
        });
    });

    describe("with multiple instances", () => {
        beforeAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                filterDispatcher: jest.fn(),
                singleAttribute: new ListAttributeValueBuilder().withType("String").withFilterable(true).build()
            } as FilterContextValue);
        });

        it("renders with a unique id", () => {
            const { asFragment: fragment1 } = render(<DatagridTextFilter {...commonProps} />);
            const { asFragment: fragment2 } = render(<DatagridTextFilter {...commonProps} />);

            expect(fragment1().querySelector("button")?.getAttribute("aria-controls")).not.toBe(
                fragment2().querySelector("button")?.getAttribute("aria-controls")
            );
        });

        afterAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });
    });

    describe("events", () => {
        let dispatch: jest.Mock;
        let parentChannelName: string;
        let ctx: FilterContextValue;
        beforeEach(() => {
            dispatch = jest.fn();
            parentChannelName = Math.random().toString(36).slice(-10);
            ctx = {
                filterDispatcher: dispatch,
                eventsChannelName: parentChannelName,
                singleAttribute: new ListAttributeValueBuilder().withType("String").withFilterable(true).build()
            };
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext(ctx);
            deletePlugin();
        });

        it("resets value on external event", async () => {
            const plugin = requirePlugin();

            expect(dispatch).toHaveBeenCalledTimes(0);

            render(<DatagridTextFilter {...commonProps} defaultValue={dynamicValue<string>("foo")} name="widget_x" />);

            const input = screen.getByRole("textbox");
            expect(dispatch).toHaveBeenCalledTimes(1);
            expect(input).toHaveValue("foo");

            act(() => plugin.emit("widget_x", "reset.value"));

            expect(dispatch).toHaveBeenCalledTimes(2);
            const [{ getFilterCondition }] = dispatch.mock.lastCall;
            expect(input).toHaveValue("");
            expect(getFilterCondition()).toEqual(undefined);
        });

        it("resets value on parent event", async () => {
            const plugin = requirePlugin();

            expect(dispatch).toHaveBeenCalledTimes(0);

            render(<DatagridTextFilter {...commonProps} defaultValue={dynamicValue<string>("bar")} name="widget_x" />);

            const input = screen.getByRole("textbox");
            expect(dispatch).toHaveBeenCalledTimes(1);
            expect(input).toHaveValue("bar");

            act(() => plugin.emit(parentChannelName, "reset.value"));

            expect(dispatch).toHaveBeenCalledTimes(2);
            const [{ getFilterCondition }] = dispatch.mock.lastCall;
            expect(input).toHaveValue("");
            expect(getFilterCondition()).toEqual(undefined);
        });
    });
});
