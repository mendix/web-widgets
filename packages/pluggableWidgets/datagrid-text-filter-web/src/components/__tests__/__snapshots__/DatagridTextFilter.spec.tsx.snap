// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Filter with single instance with multiple attributes renders correctly 1`] = `
<DocumentFragment>
  <div
    class="filter-container filter-custom-class"
    data-focusindex="0"
  >
    <div
      class="filter-selector"
    >
      <div
        class="filter-selector-content"
      >
        <button
          aria-controls="TextFilter1-filter-selectors"
          aria-expanded="false"
          aria-haspopup="true"
          class="btn btn-default filter-selector-button button-icon equal"
        >
           
        </button>
      </div>
    </div>
    <input
      class="form-control filter-input"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Text Filter with single instance with no context renders error message 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger"
  >
    The Text filter widget must be placed inside the header of the Data grid 2.0 or Gallery widget.
  </div>
</DocumentFragment>
`;

exports[`Text Filter with single instance with single attribute renders correctly 1`] = `
<DocumentFragment>
  <div
    class="filter-container filter-custom-class"
    data-focusindex="0"
  >
    <div
      class="filter-selector"
    >
      <div
        class="filter-selector-content"
      >
        <button
          aria-controls="TextFilter1-filter-selectors"
          aria-expanded="false"
          aria-haspopup="true"
          class="btn btn-default filter-selector-button button-icon equal"
        >
           
        </button>
      </div>
    </div>
    <input
      class="form-control filter-input"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Text Filter with single instance with wrong attribute's type renders error message 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger"
  >
    The attribute type being used for Text filter is not 'Hashed string or String'
  </div>
</DocumentFragment>
`;

exports[`Text Filter with single instance with wrong multiple attributes' types renders error message 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger"
  >
    The Text filter widget can't be used with the filters options you have selected. It requires a "Hashed string or String" attribute to be selected.
  </div>
</DocumentFragment>
`;
